src/backends/policy_engine/souffle/datalog_lowering_visitor_test.cc

```diff
@@ -0,0 +1,243 @@
+//-----------------------------------------------------------------------------
+// Copyright 2022 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//----------------------------------------------------------------------------
+
+#include "datalog_lowering_visitor.h"
+
+#include "absl/strings/string_view.h"
+#include "src/common/testing/gtest.h"
+#include "src/ir/block_builder.h"
+#include "src/ir/datalog/value.h"
+#include "src/ir/module.h"
+
+namespace raksha::backends::policy_engine::souffle {
+
+using testing::Combine;
+using testing::TestWithParam;
+using testing::ValuesIn;
+
+struct IrAndDatalogOperationPairs {
+  const ir::Operation *ir_operation;
+  const ir::datalog::IsOperationFact *datalog_is_operation;
+};
+
+class IrOperationLoweringTest
+    : public TestWithParam<IrAndDatalogOperationPairs> {};
+
+// Check that the IrOperation is lowered to the expected `IsOperationFact`.
+// Rather than introducing a `operator==` for this, we just rely on them
+// producing the same `ToDatalogString` result.
+TEST_P(IrOperationLoweringTest, IrOperationLoweringTest) {
+  const auto [ir_operation, datalog_is_operation] = GetParam();
+
+  ir::SsaNames ssa_names;
+  DatalogLoweringVisitor datalog_lowering_visitor(ssa_names);
+  ir_operation->Accept(datalog_lowering_visitor);
+  const RakshaDatalogFacts &facts = datalog_lowering_visitor.datalog_facts();
+  EXPECT_EQ(facts.ToDatalogString(), datalog_is_operation->ToDatalogString());
+}
+
+static const ir::Operator kLiteralOperator("sql.ReadLiteral");
+
+static const ir::Operation kStringLiteralIrOperation(
+    nullptr, kLiteralOperator,
+    {{"literal", ir::Attribute::Create<ir::StringAttribute>("number_5")}}, {});
+
+static const ir::datalog::IsOperationFact kStringLiteralDatalogIsOperation(
+    ir::datalog::Operation(
+        ir::datalog::Symbol("UNKNOWN"), ir::datalog::Symbol("sql.ReadLiteral"),
+        ir::datalog::Symbol("%0.out"), ir::datalog::OperandList(),
+        ir::datalog::AttributeList(
+            ir::datalog::Attribute(ir::datalog::Symbol("literal"),
+                                   ir::datalog::StringAttributePayload(
+                                       ir::datalog::Symbol("number_5"))),
+            ir::datalog::AttributeList())));
+
+static const ir::Operator kMergeOpOperator("sql.MergeOp");
```

Comment:
Why are we not using the factory method in https://github.com/google-research/raksha/blob/main/src/frontends/sql/ops/merge_op.h?

(Given that tests also act as documentation, it is better to use the right API here.)

Reply:
The factory method requires an `IRContext` that has been pre-populated with the appropriate `Operator`s. The interface to `IRContext` makes it hard to populate that in a `const` global. So I'm just creating some `Operator`s on the fly, as that is more concise.

Reply:
Why can't we just define all of this in a class and use it? I am mainly pushing back because test act as documentation and don't want to leave the wrong usage here. 

See https://github.com/google-research/raksha/blob/eee95e5c52f186b44674f0379bc1c1e397ad29ab/src/frontends/sql/ops/merge_op_test.cc#L37 and https://github.com/google-research/raksha/blob/eee95e5c52f186b44674f0379bc1c1e397ad29ab/src/frontends/sql/ops/tag_transform_op_test.cc#L35

We should probably create a test util class, but we can defer that for later.

Reply:
So, the issue is that these are used in constructing data parameter objects, which must be fully formed before any test or test fixture code runs. Currently, we have to fill an `IRContext` with the structures that it tracks via `RegisterOperator` and/or `RegisterStorage` function calls from test or test fixture code.

It doesn't necessarily have to be that way. It is easy to imagine adding a constructor onto `IRContext` taking `std::vector`s of `Operator`s and `Storage`s and statically seeding the context with those objects. This would allow a `const` global `IRContext` to be created statically, which would make it possible to construct the `IRContext` before any of these `Operation`s and would solve this problem. But that requires some not hard but not trivial refactoring in `IRContext`.

So I think this is, unfortunately, the best we can do for now, but I think we can do better with an additional constructor in `IRContext`.

Reply:
See #540, which contains the change that I think would be necessary to make using `IRContext` in this fashion easier.

Reply:
Actually, I just realized that we should be testing for operations in general (and not for sql operators). As long as we test operators that have  inputs, attributes, and outputs, we are fine. .